core/src/bytes/zarith/zarith.h

```diff
@@ -0,0 +1,160 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+
+#ifndef LEDGER_CORE_ZARITH_H
+#define LEDGER_CORE_ZARITH_H
+
+#include <algorithm>
+#include <vector>
+#include <bytes/BytesReader.h>
+#include <utils/hex.h>
+
+namespace ledger {
+    namespace core {
+        class zarith {
+        public:
+            /*
+             * Most significant bit of each byte is replaced with 1 to let
+             * us know there is still something to read, else 0
+             * To sum up, if the resulting bytes are "chained" together (little endian) and we drop
+             * the most significant bit of each byte, we get the input ("chained") bytes
+             * e.g.
+             * Input:           ad d9                  10101101 11011001  (I)
+             * Output:          d9 db 02      11011001 11011011 00000010
+             * LE:              02 db d9      00000010 11011011 11011001
+             * Drop 1st bit:            (0)0000010 (1)1011011 (1)1011001  == (I)
+            */
+            static std::vector<uint8_t> zSerializeNumber(const std::vector<uint8_t> &inputData) {
+                std::vector<uint8_t> result;
+                size_t id = inputData.size() - 1, delay = 0; // delay is used when offset > 7
+                uint8_t offset = 0;
+                while (id + delay >= 0 && id + delay < inputData.size()) {
+                    auto byte = inputData[id + delay];
+
+                    // We shift current byte
+                    byte <<= offset;
+
+                    // Take shifted bits from previous byte
+                    byte |= id + delay < inputData.size() - 1 ? inputData[id + delay + 1] >> (7 - (offset - 1)) : 0x00;
+
+                    // This boolean will let us know if shifting last byte of inputData
+                    // will generate a new byte (with remaining shifted bytes)
+                    bool isFirst = id + delay == 0;
+                    bool needAdditionalByte = isFirst && inputData[id + delay] >> (7 - offset);
+
+                    // We set first bit to 1 to know that there is a coming byte
+                    if (!isFirst || needAdditionalByte) {
+                        byte |= 0x80;
+                    }
+
+                    // Push to result modified byte
+                    result.push_back(byte);
+                    if (needAdditionalByte) {
+                        result.push_back(inputData[id + delay] >> (7 - offset));
+                    }
+
+                    // Update offset: if offset % 7 == 0 then following byte
+                    // is not shifted i.e. offset = 0
+                    offset = offset > 0 && !(offset % 7) ? 0 : offset + 1;
```

Comment:
`offset = (offset + 1) % 7;` is more clear IMO

Reply:
It's not the equivalent (e.g. offset = 7)

Reply:
Ah, I meant `offset = (offset +1) % 8;`